### PR TITLE
[do not merge] Ban `std::env::{set_var, remove_var}` in unsafe

### PR DESCRIPTION
**Pull request by jhpratt**
This PR prohibits any use of `std::env::ser_var` and `std::env::remove_var` in an unsafe context. It is an experiment to see how many times this actually occurs in real-world code. As such, it should not be merged.

cc #90308

@rustbot label -A-testsuite

### CODE REVIEW

**Comment by jhpratt** [comment]
If anyone on @rust-lang/compiler knows how to avoid this issue, it would be great ^^ I added two new lang items (behind `#[cfg(not(bootstrap))]` and am getting that error. The overall diff is quite small.

File: library/std/src/env.rs [hunk 1]
```diff
@@ -343,6 +343,7 @@ impl Error for VarError {
 /// env::set_var(key, "VALUE");
 /// assert_eq!(env::var(key), Ok("VALUE".to_string()));
 /// ```
+#[cfg_attr(not(bootstrap), lang = "env_set_var")]
```
Comment (lukas-code) [inline library/std/src/env.rs:346]:
```suggestion
#[cfg_attr(all(not(bootstrap), not(test)), lang = "env_set_var")]
```

File: library/std/src/env.rs [hunk 2]
```diff
@@ -385,6 +386,7 @@ fn _set_var(key: &OsStr, value: &OsStr) {
 /// env::remove_var(key);
 /// assert!(env::var(key).is_err());
 /// ```
+#[cfg_attr(not(bootstrap), lang = "env_remove_var")]
```
Comment (lukas-code) [inline library/std/src/env.rs:389]:
```suggestion
#[cfg_attr(all(not(bootstrap), not(test)), lang = "env_remove_var")]
```

**Review by lukas-code — COMMENTED** [review]
This happens, because the lang items are duplicated in the std crate and the std [unit tests](https://github.com/rust-lang/rust/tree/master/library/std/tests). The fix should be to disable the lang items for the tests.

**Comment by briansmith** [comment]
What if the user pre-emptively put these calls in `unsafe { }` blocks before the functions were marked as `unsafe`? That's exactly the right thing to do, isn't it?

**Comment by jhpratt** [comment]
@briansmith _Hopefully_ there aren't too many places where this errors, so that the ones that do can be manually reviewed. Sure, there are probably a couple that have preemptively placed it in unsafe blocks, but my suspicion is that in those situations it's the _only_ thing in an unsafe block.

**Comment by jhpratt** [comment]
Also, CI is now green, so a crater run can be kicked off. The queue is currently empty.

**Comment by RalfJung** [comment]
> What if the user pre-emptively put these calls in unsafe { } blocks before the functions were marked as unsafe? That's exactly the right thing to do, isn't it?

Seems premature to do that given that we haven't documented the safety requirements of those functions yet...

**Comment by oli-obk** [comment]
@craterbot check

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-103752/retry-regressed-list.txt

**Comment by RalfJung** [comment]
True regressions (deduplicated):
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/aorist_extendr-engine-0.0.1/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/bigbro-0.5.2/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/extendr-engine-0.3.1/log.txt (basically same as the first but seems to be a copy of that code)
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/libafl-0.8.2/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/gh/43615.dcim/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/gh/YangKeao.mntexec/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/auxv-0.3.3/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/cassandra-cpp-sys-1.1.0/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/cql_bindgen-0.2.0/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/libR-sys-0.3.0/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/nstd_env-0.5.0/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/s2n-tls-sys-0.0.17/log.txt
- https://crater-reports.s3.amazonaws.com/pr-103752-1/try%233aa941b8e3d5c6f0f1ab19a18debdc2dfe4c87bb/reg/s3-batch-put-tar-0.1.0/log.txt

So, there are some crates out there that do env accesses from unsafe code and thus would not be warned at all by a `deprecated_safe`. Not a huge number but still.

**Comment by jhpratt** [comment]
Closing as crater has been run.

**Pull request closed without merging**